src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1041,7 +1051,13 @@ public static List<TSource> ToList<TSource>(this IEnumerable<TSource> source)
             if (source == null) throw Error.ArgumentNull("source");
             if (keySelector == null) throw Error.ArgumentNull("keySelector");
             if (elementSelector == null) throw Error.ArgumentNull("elementSelector");
-            Dictionary<TKey, TElement> d = new Dictionary<TKey, TElement>(comparer);
+
+            int capacity = 0;
+
+            if (source is TSource[]) capacity = (source as TSource[]).Length; // Cast to Array is much faster than cast to ICollection
+            else if (source is ICollection<TSource>) capacity = (source as ICollection<TSource>).Count;
```

Comment:
Consider refactoring out these checks into a helper so that we don't duplicate them in both functions.